component/amqp/component.go

```diff
@@ -0,0 +1,359 @@
+	defer tickerStats.Stop()
+
+	btc := &batch{
+		ctx:      ctx,
+		messages: make([]Message, 0, c.batchCfg.count),
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			log.Info("context cancellation received. exiting...")
+			return ctx.Err()
+		case delivery, ok := <-sub.deliveries:
+			if !ok {
+				return errors.New("subscription channel closed")
+			}
+			log.Debugf("processing message %d", delivery.DeliveryTag)
+			observeReceivedMessageStats(c.queue, delivery.Timestamp)
+			c.processBatch(c.createMessage(ctx, delivery), btc)
+		case <-batchTimeout.C:
+			log.Debugf("batch timeout expired, sending batch")
+			c.sendBatch(btc)
+		case <-tickerStats.C:
+			err := c.stats(sub)
+			if err != nil {
+				log.Errorf("failed to report sqsAPI stats: %v", err)
+			}
+		}
+	}
+}
+
+func observeReceivedMessageStats(queue string, timestamp time.Time) {
+	messageAge.WithLabelValues(queue).Set(time.Now().UTC().Sub(timestamp).Seconds())
+	messageCountInc(queue, fetchedMessageState, nil)
+}
+
+type subscription struct {
+	conn       *amqp.Connection
+	channel    *amqp.Channel
+	deliveries <-chan amqp.Delivery
+	closed     bool
+}
+
+func (s *subscription) close() error {
+	if s.closed {
+		return nil
+	}
+	var ee []error
+	if s.channel != nil {
+		ee = append(ee, s.channel.Close())
+	}
+	if s.conn != nil {
+		ee = append(ee, s.conn.Close())
+	}
+	s.closed = true
+	return patronerrors.Aggregate(ee...)
+}
+
+func (c *Component) subscribe() (subscription, error) {
+	conn, err := amqp.DialConfig(c.url, c.cfg)
+	if err != nil {
+		return subscription{}, fmt.Errorf("failed to dial @ %s: %w", c.url, err)
+	}
+	sub := subscription{conn: conn}
+
+	ch, err := conn.Channel()
+	if err != nil {
+		return subscription{}, patronerrors.Aggregate(conn.Close(), fmt.Errorf("failed get channel: %w", err))
+	}
+	sub.channel = ch
+
+	tag := uuid.New().String()
+	log.Infof("consuming messages for tag %s", tag)
+
+	deliveries, err := ch.Consume(c.queue, tag, false, false, false, false, nil)
+	if err != nil {
+		return subscription{}, patronerrors.Aggregate(ch.Close(), conn.Close(), fmt.Errorf("failed initialize amqp consumer: %w", err))
+	}
+	sub.deliveries = deliveries
+
+	return sub, nil
+}
+
+func (c *Component) createMessage(ctx context.Context, delivery amqp.Delivery) *message {
+	corID := getCorrelationID(delivery.Headers)
+	sp, ctxMsg := trace.ConsumerSpan(ctx, trace.ComponentOpName(consumerComponent, c.queue),
+		consumerComponent, corID, mapHeader(delivery.Headers), c.traceTag)
+
+	ctxMsg = correlation.ContextWithID(ctxMsg, corID)
+	ctxMsg = log.WithContext(ctxMsg, log.Sub(map[string]interface{}{correlation.ID: corID}))
+
+	return &message{
+		ctx:     ctxMsg,
+		span:    sp,
+		msg:     delivery,
+		requeue: c.requeue,
+		queue:   c.queue,
+	}
+}
+
+func (c *Component) processBatch(msg *message, btc *batch) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	btc.append(msg)
+
+	if len(btc.messages) >= int(c.batchCfg.count) {
+		c.proc(btc)
+		btc.reset()
```

Comment:
maybe a very small improvement might be to extract these 2 common lines and re-use right below. And also maybe add some metric for how many times which batch trigger was used (time or count)

Reply:
Introduced a method as suggested but I am unsure about the metric. What do you want us to measure?

Reply:
it s basically a measure of throughput what i had in mind ... (number of events) / (time interval) . So by tracking `how many times we did a batch reset due to time` and `how many times we did a batch reset due to message count` . you get a throughput index "for free" . It s not very important, but might be interesting to track. It might also point to some configuration adjustments (adjust time interval for example to ensure reasonable batch size) ... 
Might not be the best idea, but i was wondering if this would make sense also to others as a valuable metric.

Reply:
Since we are measuring individual message processing metrics I believe we should be covered.
Would it make more sense to measure how many batches we are processing?
In case batch size == 1 then message an batch processing will be the same.